pynumaflow/function/multiproc_server.py

```diff
@@ -0,0 +1,260 @@
+                    watermark=request.watermark.watermark.ToDatetime(),
+                    metadata=DatumMetadata(
+                        msg_id=request.metadata.id,
+                        num_delivered=request.metadata.num_delivered,
+                    ),
+                ),
+            )
+        except Exception as err:
+            _LOGGER.critical("UDFError, re-raising the error: %r", err, exc_info=True)
+            raise err
+
+        datums = []
+        for msgt in msgts.items():
+            event_time_timestamp = _timestamp_pb2.Timestamp()
+            event_time_timestamp.FromDatetime(dt=msgt.event_time)
+            watermark_timestamp = _timestamp_pb2.Timestamp()
+            watermark_timestamp.FromDatetime(dt=request.watermark.watermark.ToDatetime())
+            datums.append(
+                udfunction_pb2.DatumResponse(
+                    keys=list(msgt.keys),
+                    value=msgt.value,
+                    tags=msgt.tags,
+                    event_time=udfunction_pb2.EventTime(event_time=event_time_timestamp),
+                    watermark=udfunction_pb2.Watermark(watermark=watermark_timestamp),
+                )
+            )
+        return udfunction_pb2.DatumResponseList(elements=datums)
+
+    # def ReduceFn(
+    #         self,
+    #         request_iterator: AsyncIterable[udfunction_pb2.Datum],
+    #         context: NumaflowServicerContext,
+    # ) -> udfunction_pb2.DatumList:
+    #     """
+    #     Applies a reduce function to a datum stream.
+    #     The pascal case function name comes from the proto udfunction_pb2_grpc.py file.
+    #     """
+    #     _LOGGER.error("Reduce not supported on NEW SYNC --")
+    #     raise NotImplementedError("Reduce Not supported on sync")
+
+    def IsReady(
+        self, request: _empty_pb2.Empty, context: NumaflowServicerContext
+    ) -> udfunction_pb2.ReadyResponse:
+        """
+        IsReady is the heartbeat endpoint for gRPC.
+        The pascal case function name comes from the proto udfunction_pb2_grpc.py file.
+        """
+        return udfunction_pb2.ReadyResponse(ready=True)
+
+    def _run_server(self, bind_address):
+        """Start a server in a subprocess."""
+        _LOGGER.info("Starting new server.")
+        options = [("grpc.so_reuseport", 1), ("grpc.so_reuseaddr", 1)]
+        for x in options:
+            self._server_options.append(x)
+        server = grpc.server(
+            futures.ThreadPoolExecutor(
+                max_workers=self._THREAD_CONCURRENCY,
+            ),
+            options=self._server_options,
+        )
+        udfunction_pb2_grpc.add_UserDefinedFunctionServicer_to_server(self, server)
+        server.add_insecure_port(bind_address)
+        server.start()
+        _LOGGER.info("GRPC Multi-Processor Server listening on: %s %d", bind_address, os.getpid())
+        server.wait_for_termination()
+
+    @contextlib.contextmanager
+    def _reserve_port(self) -> int:
```

Comment:
See if you can use the `Generator` typing here. Since you are not returning, but actually yielding